src/probnum/diffeq/_perturbsolve_ivp.py

```diff
@@ -0,0 +1,191 @@
+"""Perturbation-based probabilistic ODE solver."""
+import numpy as np
+import scipy.integrate
+
+from probnum import problems
+from probnum.diffeq import perturbed, stepsize
+
+__all__ = ["perturbsolve_ivp"]
+
+
+SELECT_METHOD = {
+    "RK45": scipy.integrate.RK45,
+    "RK23": scipy.integrate.RK23,
+}
+METHODS = SELECT_METHOD.keys()
+
+SELECT_PERTURBS = {
+    "step-lognormal": perturbed.step.PerturbedStepSolver.construct_with_lognormal_perturbation,
+    "step-uniform": perturbed.step.PerturbedStepSolver.construct_with_uniform_perturbation,
+}
+PERTURBS = SELECT_PERTURBS.keys()
+
+
+def perturbsolve_ivp(
+    f,
+    t0,
+    tmax,
+    y0,
+    rng,
+    method="RK45",
+    perturb="step-uniform",
+    noise_scale=10.0,
+    adaptive=True,
+    atol=1e-2,
+    rtol=1e-2,
+    step=None,
+):
+    r"""Solve an initial value problem with a perturbation-based probabilistic ODE solver.
+
+    Parameters
+    ----------
+    f :
+        ODE vector field.
+    t0 :
+        Initial time point.
+    tmax :
+        Final time point.
+    y0 :
+        Initial value.
+    rng :
+        Random number generator.
+    method :
+        Integration method to use.
+        The following are available (docs adapted from `SciPy <https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html>`_):
+
+            * `RK45` (default): Explicit Runge-Kutta method of order 5(4) [2]_.
+              The error is controlled assuming accuracy of the fourth-order
+              method, but steps are taken using the fifth-order accurate
+              formula (local extrapolation is done). A quartic interpolation
+              polynomial is used for the dense output [3]_. Can be applied in
+              the complex domain.
+            * `RK23`: Explicit Runge-Kutta method of order 3(2) [4]_. The error
+              is controlled assuming accuracy of the second-order method, but
+              steps are taken using the third-order accurate formula (local
+              extrapolation is done). A cubic Hermite polynomial is used for the
+              dense output. Can be applied in the complex domain.
+
+        Other integrators are not supported currently.
+    perturb
+        Which perturbation style to use.
+        Currently, the following methods are supported:
+
+            * `step-lognormal`: Perturbed-step (aka random time-step numerical integration) method
+              with lognormally distributed perturbation of the step-size [1]_.
+            * `step-uniform`: Perturbed-step (aka random time-step numerical integration) method
+              with lognormally distributed perturbation of the step-size [1]_.
+
+        Other integrators are not supported currently.
+    noise_scale
+        Scale of the perturbation. Optional. Default is 10.0. The magnitude of this parameter
+        significantly impacts the width of the posterior.
+    adaptive :
+        Whether to use adaptive steps or not. Default is `True`.
```

Comment:
Interesting! I somehow thought that step-size adaptation is kind of an unsolved thing with these methods? Or was that only for the Conrad method?

Reply:
I dont know where you got that from... it works :) Maybe it is not implemented elsewhere (I guess you are referring to whatever DifferentialEquations.jl does for UQ), because it is really hard to batch, or do dense output or whatever. I dont know.

Reply:
DifferentialEquations.jl actually has an adaptive-step version for the Conrad solver that basically selects the diffusion from the local error estimate and some convergence rate, but that does not have a proper reference I think. Maybe I mixed the "how to calibrate" and "how to select step sizes" part, so I guess I have to check out the paper again to see if/how they do things.